examples/move/big_vector/sources/big_vector.move

```diff
@@ -0,0 +1,1274 @@
+    const EBadRemove: u64 = 7;
+
+    /// Found a pair of nodes that are expected to be adjacent but
+    /// whose linked list pointers don't match up.
+    const ENotAdjacent: u64 = 8;
+
+    /// Tried to redistribute between two nodes, but the operation
+    /// would have had no effect.
+    const EBadRedistribution: u64 = 9;
+
+    // === Constants ===
+
+    /// Sentinel representing the absence of a slice.
+    const NO_SLICE: u64 = 0;
+
+    /// We will accommodate at least this much fan out before
+    /// splitting interior nodes, so that after the split, we don't
+    /// get an interior node that contains only one child.
+    const MIN_FAN_OUT: u64 = 4;
+
+    /// Internal nodes of `BigVector` can't have more children than
+    /// this, to avoid hitting object size limits.
+    const MAX_FAN_OUT: u64 = 4096;
+
+    /// Leaf nodes of `BigVector` can't be bigger than this, to avoid
+    /// hitting object size limits.
+    const MAX_SLICE_SIZE: u64 = 256 * 1024;
+
+    // === Removal fix-up strategies ===
+
+    /// 0b000: No fix-up.
+    const RM_FIX_NOTHING: u8 = 0;
+
+    /// 0b001: Node is completely empty (applies only to root).
+    const RM_FIX_EMPTY: u8 = 1;
+
+    /// 0b010: Stole a key from the left neighbour, additional value
+    /// is the new pivot after the steal.
+    const RM_FIX_STEAL_L: u8 = 2;
+
+    /// 0b011: Stole a key from the right neighbour, additional value
+    /// is the new pivot after the steal.
+    const RM_FIX_STEAL_R: u8 = 3;
+
+    /// 0b100: Merged with the left neighbour.
+    const RM_FIX_MERGE_L: u8 = 4;
+
+    /// 0b101: Merged with the right neighbour.
+    const RM_FIX_MERGE_R: u8 = 5;
+
+    // === Constructors ===
+
+    /// Construct a new, empty `BigVector`. `max_slice_size` contains
+    /// the maximum size of its leaf nodes, and `max_fan_out` contains
+    /// the maximum fan-out of its interior nodes.
+    public fun empty<E: store>(
+        max_slice_size: u64,
+        max_fan_out: u64,
+        ctx: &mut TxContext,
+    ): BigVector<E> {
+        assert!(0 < max_slice_size, ESliceTooSmall);
+        assert!(max_slice_size <= MAX_SLICE_SIZE, ESliceTooBig);
+        assert!(MIN_FAN_OUT <= max_fan_out, EFanOutTooSmall);
+        assert!(max_fan_out <= MAX_FAN_OUT, EFanOutTooBig);
+
+        BigVector {
+            id: ctx.new(),
+
+            depth: 0,
+            length: 0,
+
+            max_slice_size,
+            max_fan_out,
+
+            root_id: NO_SLICE,
+            last_id: NO_SLICE,
+        }
+    }
+
+    /// Destroy `self` as long as it is empty, even if its elements
+    /// are not droppable. Fails if `self` is not empty.
+    public fun destroy_empty<E: store>(self: BigVector<E>) {
+        let BigVector {
+            id,
+
+            depth: _,
+            length,
+            max_slice_size: _,
+            max_fan_out: _,
+
+            root_id: _,
+            last_id: _,
+        } = self;
+
+        assert!(length == 0, ENotEmpty);
```

Comment:
I tend not to worry about the ordering of these things because the assertion failure is game over -- nothing that happened before or after it will have any effect.